admin/autoupdater/src/autoupdater.c

```diff
@@ -0,0 +1,458 @@
+		fprintf(stderr, "autoupdater: warning: no matching firmware found (model %s)\n", platforminfo_get_image_name());
+		return false;
+	}
+
+	/* Check manifest signatures */
+	{
+		ecc_int256_t hash;
+		ecdsa_sha256_final(&m->hash_ctx, hash.p);
+		ecdsa_verify_context_t ctxs[m->n_signatures];
+		for (size_t i = 0; i < m->n_signatures; i++)
+			ecdsa_verify_prepare_legacy(&ctxs[i], &hash, m->signatures[i]);
+
+		long unsigned int good_signatures = ecdsa_verify_list_legacy(ctxs, m->n_signatures, s->pubkeys, s->n_pubkeys);
+		if (good_signatures < s->good_signatures) {
+			fprintf(stderr, "autoupdater: warning: manifest %s only carried %lu valid signatures, %lu are required\n", manifest_url, good_signatures, s->good_signatures);
+			return false;
+		}
+	}
+
+	/* Check version and update probability */
+	if (!newer_than(m->version, s->old_version)) {
+		puts("No new firmware available.");
+		return true;
+	}
+
+	if (!s->force && random() >= RAND_MAX * get_probability(m->date, m->priority, s->fallback)) {
+		fputs("autoupdater: info: no autoupdate this time. Use -f to override.\n", stderr);
+		return true;
+	}
+
+	/**** Download and verify image file *********************************/
+	/* Begin download of the image */
+	run_dir(download_d_dir);
+
+	struct recv_image_ctx image_ctx = { };
+	image_ctx.fd = open(firmware_path, O_WRONLY|O_CREAT, 0600);
+	if (image_ctx.fd < 0) {
+		fprintf(stderr, "autoupdater: error: failed opening firmware file %s\n", firmware_path);
+		goto fail_after_download;
+	}
+
+	/* Download image and calculate SHA256 checksum */
+	{
+		char image_url[strlen(mirror) + strlen(m->image_filename) + 2];
+		sprintf(image_url, "%s/%s", mirror, m->image_filename);
+		ecdsa_sha256_init(&image_ctx.hash_ctx);
+		int err_code = get_url(image_url, &recv_image_cb, &image_ctx);
+		if (err_code != 0) {
+			fprintf(stderr, "autoupdater: warning: error downloading image: %s\n", uclient_get_errmsg(err_code));
+			close(image_ctx.fd);
+			goto fail_after_download;
+		}
+	}
+	close(image_ctx.fd);
+
+	/* Verify image checksum */
+	{
+		ecc_int256_t hash;
+		ecdsa_sha256_final(&image_ctx.hash_ctx, hash.p);
+		if (memcmp(hash.p, m->image_hash, ECDSA_SHA256_HASH_SIZE)) {
+			fputs("autoupdater: warning: invalid image checksum!\n", stderr);
+			goto fail_after_download;
+		}
+	}
+
+	/**** Call sysupgrade ************************************************/
+	if (s->simulate) {
+		printf(
+			"autoupdater: info: Aborting successful upgrade because simulation was requested.\n"
+			"autoupdater: info: You can find the firmware file in %s\n",
+			firmware_path
+		);
+		run_dir(abort_d_dir);
+		return true;
+	}
+
+	/* Begin upgrade */
+	run_dir(upgrade_d_dir);
+
+	/* Close stdin and stderr to not give sysupgrade unneccessary SIGHUPs */
+	{
+		int null_fd = open("/dev/null", O_RDWR);
+		dup2(null_fd, 0);
+		dup2(null_fd, 2);
+		close(null_fd);
+	}
+
+	/* Unset FD_CLOEXEC so the lockfile stays locked during sysupgrade */
+	fcntl(lock_fd, F_SETFD, 0);
```

Comment:
But does it harm anybody? Maybe people want to backport the autoupdater to earlier OpenWrts or call an entirely different program for whatever reason. It feels cleaner to let the lockfile stay open across the exec.